### ISSUE TITLE
new GPO operator .map

### ISSUE THREAD

**Issue by p0W3RH311** [PowerShell/PowerShell#15153]
in PS there are many operator collections: like `-contains, -join, where, foreach`.... every operator performs one specific task very well (except where and foreach _GPO "General Purpose Operator"_). it's great but 
since ps does not yet implement **linq** it lacks other important operators like: `.map() `

`$collection.map( {code} , arg1, arg2, $collection2 )`

`.map()`  traverses two lists in parallel, applying a scriptblock to the corresponding items from each list. If one list is longer, the extra elements are ignored. 

```
arg1 = name of first variable in first collection
arg2 = name of second variable in second collection
```

proof of concept:
```

PS> (3,5,6).map( {  $a + $b } , a, b, (1,2,4) )
4
7
10
```

**Comment by mklement0** [comment]
I like the idea in principle, but to me a more PowerShell-idiomatic implementation would be the `foreach` statement extension I've proposed in #14732 (not limited to 2 collections):

```powershell
# WISHFUL THINKING
PS> foreach ($a, $b in (3, 5, 6), (1, 2, 4)) { $a + $b }
4
7
10
```

**Comment by p0W3RH311** [comment]
@mklement0 

thanks, +1 for your idea i like and i like more if it implement variable indexer. 
But other programmers who are used other scripting/programming languages like python/scala/haskell...etc  ​​use  traditonally statement like `map() zip() zipwith() zipwithindex()` ...etc. its awesome if PS implement statement/operators like this.

**Comment by mklement0** [comment]
Thanks, @p0W3RH311, and, yes, supporting an automatically maintained index variable, as you've proposed in #14724, is also a good idea (though in my mind it calls for an explicit opt-in with a self-chosen index variable).

As for `map() zip() zipwith() zipwithindex()`: To me (I know not everyone agrees), PowerShell-idiomatic solutions should be limited to cmdlets, operators, and language statements (such as `foreach`): _methods_ are an awkward fit, primarily due to the potential for syntax confusion (argument vs. expression mode).

However, given that LINQ methods _are_ already accessible - as something that PowerShell _provides access to_ rather than being a native part of its language, the more promising approach would be to allow LINQ methods to be called more naturally, which would require PowerShell to support _extension methods_.

Currently, using `Linq.Zip()` is awkward:

```powershell
PS> [System.Linq.Enumerable]::Zip([int[]] (3, 5, 6), [int[]] (1, 2, 4), [Func[int, int, int]] { $args[0] + $args[1] })
4
7
10
```

I think improving this experience is the more promising approach to pursue - see #2226.

---

As for `zipwithindex()`, specifically: my thinking behind  #14732 was to allow it to be _combined_ with #14724.

```powershell
# WISHFUL THINKING
PS> foreach ($a, $b in (3, 5, 6), (1, 2, 4); $ndx) { "$ndx`:" + ($a + $b) }
0: 4
1: 7
2: 10
```

**Comment by p0W3RH311** [comment]
with native supporting `Linq ` language PS will become `PS++` :) ... all scripting/programming languages have an orientation to functional paradigm and i believe its the time to PS to to follow this path...the best solution is `Linq`

**Comment by iRon7** [comment]
Looking to some recent proposals and requests for parallel lists and a iterator index, there is clearly a need to join objects side-by-side. Therefore, I did a propose to [**add a `Join-Object` cmdlet to the standard PowerShell equipment `#14994`**](https://github.com/PowerShell/PowerShell/issues/14994).
In this idea (and existing example), the [`Join-Object`](https://www.powershellgallery.com/packages/Join) cmdlet has a brought scalar of join features from **joining powershell objects based a related property** to **side-by-side joining (traversing) scalars (as primitives and strings)**:


#### Example 1
```PowerShell
# traverse two lists
3, 5, 6 |Join 1, 2, 4 |% { "$_" }
3 1
5 2
6 4
```

#### Example 2
```PowerShell
# Sum each item in two parallel lists
3, 5, 6 |Join 1, 2, 4 |% { $_[0] + $_[1] }
4
7
10
```

#### Example 3
```PowerShell
# Sum each item in multiple parallel lists
3, 5, 6 |Join 1, 2, 4 |Join 7, 8, 9 |% { ($_ |Measure-Object -sum).sum }
11
15
19
```

#### Example 4
```PowerShell
# Index and sum each item in the additional parallel lists
0..9 |Join 3, 5, 6 |Join 1, 2, 4 |% { "$($_[0]): " + ($_[1] + $_[2]) }
0: 4
1: 7
2: 10
```

#### Example 5
```PowerShell
# Same as above , but with named properties
0..9 |Join 3, 5, 6 |Join 1, 2, 4 -Name Index, a, b |% { "$($_.Index): " + ($_.a + $_.b) }
0: 4
1: 7
2: 10
```

**Comment by daxian-dbw** [comment]
Language working group:
This issue is a duplicate of #2226.
